### DISCUSSION TITLE
Usage with Ionic Capacitor

### DISCUSSION THREAD

**Discussion by leo-petrucci** [nextauthjs/next-auth#4446] [Help]
### Question 💬

This isn't as much of a question as more of a list of findings/tutorial on how to get this (kind-of) working, it's already been done for React Native #569 so hopefully it's okay for me to post this.

-----

The big issue I found while implementing this is how Android's Browser (which uses [Chrome Custom Tabs](https://developer.chrome.com/docs/android/custom-tabs/#whentouse)) handles cookies.

### Using the built in browser didn't work

Initially I was hoping I'd be able to simply open the Browser, log-in from there and then be sent back to my app with my cookies set. Any cookie saved within Chrome Custom Tabs will be saved on the systems' browser and since Capacitor does not share cookies with Chrome, that session will not be available in your Capacitor app.

### Using the Rest API almost worked

The second thing I tried was to implement the oAuth flow entirely using the [Rest API](https://next-auth.js.org/getting-started/rest-api). My theory was to:

1. Generate an oAuth URL for my provider
2. Navigate there with Chrome Custom Tabs
3. Authorise the application
4. Have a callback to a deep-link within the app so that I could get access to the generated code
5. Manually post to the Next Auth `callback` endpoint to get session data

With the API I was able to get an oAuth url to authorise my app, however the issue appeared when redirecting after authorising.

As far as I can tell the `/api/auth/signin/:provider` endpoint will generate an oAuth URL that will **always redirect to `/api/auth/callback/:provider`**. The `callbackURL` option is just a page that Next Auth will redirect you after. So there is technically no way to redirect to the app with the code.

### Workaround using the In-App Browser plugin

I was almost about to give up when I found the [Cordova In App Browser plugin](https://ionicframework.com/docs/native/in-app-browser). In App Browser essentially opens another webview on top of your app, its cookies are shared with the Capacitor webview which means any authentication done here will also work in Capacitor.

After some trial and error I ended up getting the flow to work with the following code, added comments for clarity.

```
export const loginWithTwitter = async (callback?: () => void) => {
  // Fetches the csrf token, needed for all requests
  const { csrfToken }: { csrfToken: string } = await fetch(
    `${host}/api/auth/csrf`,
    {
      credentials: "include",
    }
  ).then((res) => {
    return res.json();
  });

  // Generate an oAuth url for the Twitter provider
  const { url }: { url: string } = await fetch(
    `${host}/api/auth/signin/twitter`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        csrfToken,
        json: "true",
        // where next-auth will redirect us after auth
        callbackUrl: `${host}/redirect`,
      }),
      redirect: "follow",
      credentials: "include",
    }
  ).then((res) => res.json());

  const ref = InAppBrowser.create(
    url,
    "_blank",
    "location=yes,hidenavigationbuttons=yes,hideurlbar=yes,toolbarcolor=#f1f5f9"
  );

  if (ref) {
    /**
     * Here we check if we've been successfully redirected, if we have
     * we close the browser
     */
    ref!.on("loadstart").subscribe((e: any) => {
      const includesUrl = e.url.includes(`/redirect`);
      if (includesUrl) {
        ref.close();
      }
    });

    // The iap kind of sucks, on my Twitter login I had a bunch of elements that weren't loading
    // This injects some CSS to hide them
    ref!.on("loadstop").subscribe((e: any) => {
      if (e.url.includes(`/oauth2/authorize`)) {
        // Remove forever loading element
        ref.insertCSS({
          code: `
            div[role="progressbar"] {
              display: none;
            }
          `,
        });
      }
    });
  }
};
```

This doesn't look super legit, and since the IAP doesn't actually have access to any Cookies from Chrome, the user will be asked to login again before authorising the app.

![image](https://user-images.githubusercontent.com/23533178/161960481-8627bfb3-17d6-4fa5-b7c9-30020cf564ad.png)
![image](https://user-images.githubusercontent.com/23533178/161960737-40119797-957d-4458-a3e6-0974aae667b5.png)

That said, it does work. It's pretty much the only solution I could find short of reimplementing the entire flow.

**Note:** `SessionProvider` won't immediately have access to the session. The docs say to call `getSession` to force an update, but it didn't work for me. At the moment I'm forcing a page reload after login to make sure everything is synced properly.

Would love to know if anyone has better solutions, let me know if this has helped you.

### How to reproduce ☕️

N/A

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by rickitan** [comment]
Great finding @creativiii are you still doing it this way? By any chance did you try with Google Sign in? Will this work for iOS as well?

Thanks!

**Reply by leo-petrucci** [reply]
I'm not doing it this way anymore, I've figured out that you can use deep linking as a callback URL to automatically send users back to the app and close the "official" in-app browser.

Then from whatever page you send the user back to you can hit the next-auth API callback endpoint.

I was working on a demo but ended up losing the source code, I'm planning to start over but it might take a bit. I'll update here once it's done 👍

iOS shouldn't need any workarounds, i think the in-app browser shares cookies with the app it's opened by so you can simply have the same login flow as the web.

**Reply by rickitan** [reply]
Very cool I definitely want to see how're you doing it now. Where do you add the deep link. Is it in the in the Provider configuration page? For Google it'd be in Google console I guess. Right now I point it to my domain: `https://www.mydomain.com/api/auth/callback/google`

For your project do you use the same Next.js codebase for a web app and also the native ones?

I want to have the same codebase for Web, iOS and Android. Any good resource to learn how to achieve this? I'm new to Capacitor.

**Reply by leo-petrucci** [reply]
So you just add the link in your provider, then you should be able to use that link as a callback URL. For some reason it had to be a custom protocol deep link for me, I wouldn't work any other way (`myapp://example.com`).

> For your project do you use the same Next.js codebase for a web app and also the native ones?

Yes, Next is my API, web frontend and also my mobile frontend.

>I want to have the same codebase for Web, iOS and Android. Any good resource to learn how to achieve this? I'm new to Capacitor.

Check out this repo, should have everything you need to get started https://github.com/mlynch/nextjs-tailwind-ionic-capacitor-starter

**Reply by ryanpwaldon** [reply]
Any chance you could share a repo? I'm trying to get this working on iOS – I've got deep links working, but cookies aren't being shared :(

**Reply by choutkamartin** [reply]
@creativiii Sorry to ping you. But currently trying to solve the exact same problem and I'm asking the same thing as @ryanpwaldon. Can you share some repo, please? Thank you very much

**Reply by leo-petrucci** [reply]
Sorry guys I lost progress in the repo I was using to play around with this and I don't have time to mess with it at the moment.

Which cookies aren't being shared?

You essentially need to redirect back to your app passing the token generated by oauth, then from your app you need to hit the `callback` endpoint.

There shouldn't be cookies to share, because all requests should be done from the app itself.

**Reply by choutkamartin** [reply]
@creativiii It's fine, don't worry about it. I will create a new repo, and I will try to somehow figure it out. If I won't figure it out, or there will be some minor problem, I will maybe ping you again, if you are okay with it.

I was trying to make it work in one of my bigger projects, which was pain in the ass, so I will start from scratch, maybe that will work out haha.

**Reply by leo-petrucci** [reply]
I'm tying up some lose ends here and there on other projects, but I'll try to jump on a reproduction asap as soon as I'm done.

Let me know how it goes or if you get blocked .

**Reply by choutkamartin** [reply]
@creativiii Can you please show me how your `capacitor.config` looks? I'm Wondering if you used any extra configuration. Also, which version of  In App Browser did you use? Thanks!

**Reply by leo-petrucci** [reply]
@choutkamartin the only thing I'm seeing is that I've got my hostname set to `app.mydomain.com`, but I don't know if that was on purpose or not.

**Reply by choutkamartin** [reply]
@creativiii Surely it was on purpose. I'm doing the same. My cookies are now persisted, there's one little problem though. Cookies are not send with the session fetch that happens when `useSession()` is called. The problem is simple, but is hard to solve. The origin of the request is `app.domain.com`, however, the domain of cookie is set to `.domain.com`. The little dot before should mean that the cookie is also valid for subdomains such as `app.domain.com`, but they are simply not send with the request. When I sent a test request without `cors mode`, the cookies were sent fine. The problem is I can't send requests without `cors`. So, I will have to brainstorm for a little bit longer to make it work

**Reply by leo-petrucci** [reply]
@choutkamartin If I'm not mistaken I think I may have ended up not using `useSession` and instead making my own hook to fetch session.

**Reply by choutkamartin** [reply]
@creativiii You are right. The `SessionProvider` component, that provides `useSession` hook calls `getSession` with default `credentials: same-origin`, therefore not appending cookies to the request. Creating my custom hook with `credentials: include` fixed it. While everything works now, I think Next Auth should have an option to allow end user to decide if `getSession` should make request with `credentials: same-origin` or `credentials: include`. Thanks for help Leonardo!

**Reply by leo-petrucci** [reply]
I made an issue about it ages ago, but it never got any traction 😂

Glad you got it working!

**Reply by choutkamartin** [reply]
Yeah, Next Auth is not built so well when the API is on a different subdomain or domain. I understand that was never the use case, but it should have had this option from the start. Almost all API calls are relative to `/api/`, which is a headache as the app is trying to call itself, not the API that is running somewhere else. That's why SessionProvider has a `basePath` prop, which solves this for `/api/auth/session` requests, but it does not solve this problem for `signOut`, `signIn` methods and so on. 🙁

TLDR: There are 3-4 options:

1. Fork entire Next Auth and change what is needed
2. Write own `signIn`, `signOut`, `SessionProvider` and maybe a few more functions
3. Make a PR to Next Auth
4. Use `server.url` in Capacitor config, which I don't like, as loading entire app from my main site kind of loses the purpose of native app

**Reply by leo-petrucci** [reply]
I used `react-query` to re-implement the needed endpoints. Works quite well, hooks have invalidation and refetching features.

**Reply by BjoernRave** [reply]
@leo-petrucci could you share your solution using `react-query`?

**Reply by leo-petrucci** [reply]
> @leo-petrucci could you share your solution using `react-query`?

I don't have the code, but it's really just creating hooks that get/post to the next-auth endpoints like session and logout.

**Comment by choutkamartin** [comment]
Example repository with comments on how to make Next Auth and Capacitor work together here: https://github.com/choutkamartin/next-auth-capacitor

**Reply by datteroandrea** [reply]
If you try the live web application the authentication is not working it says "you are not signed in", also the is no example on how to make this work in android via building an APK, if needed I can provide help but I don't know where to start.

**Reply by choutkamartin** [reply]
@datteroandrea Hi! Strangely, it looks like it stopped working after I deployed it to Vercel or I just messed up something as I was testing it locally most of the time.

I'm going to check what's wrong in the upcoming days. Then I'm gonna let you know

For building the APK you need to have basic Android Studio and Capacitor understanding, I linked this article in my repository as a starter point https://devdactic.com/nextjs-and-capacitor

**Reply by datteroandrea** [reply]
@choutkamartin perfect, thank you very much, if you need any help let me know!

**Reply by choutkamartin** [reply]
It was just a small error on my side when I was refactoring it, it should be working now. I will add a downloadable APK so you can test it without building via Android Studio. Edit: I added downloadable APK as a release. Check the repository. 🙂

**Comment by gunnerhowe** [comment]
So this method won't work with iOS? Only Android?

**Comment by phil-w** [comment]
I found some of this useful - thanks - but my own solution ended up somewhat different. It feels a bit like a lash up, but it seems to work and be reliable. I'll try to summarize it here anyway , in case it helps anyone else.

I started with Sveltekit and was loathe to adopt an Ionic approach. I was a bit surprised there wasn't an "off the shelf" solution for using OpenAuth across different platforms, but it seems there isn't. I played a lot with auth.js and also Lucia, which trades complexity for flexibility. I wanted to make auth.js work as I like simple.

I started with the standard "in app browser" but I couldn't make that work, and eventually found  [an unofficial version](https://github.com/Cap-go/capacitor-inappbrowser) which worked straight away.

My problem was broadly to authenticate a sveltekit API using auth.js, such that I can authenticate using OpenAuth on web or mobile.

 - I started by making my "sign in" and "sign out" controls and calling the standard auth.js stuff when on web, but when "on capacitor", I use my own local scripts.
 - I modded my web "sign in" process so if you hit it from mobile, it directly starts the 3rd party sign in process, so you never see any of my UI in that case, just the 3rd party provider's.

### Then, when the user wants to sign in on Capacitor...
 1. My UI invokes the IAB with a path which is eg `http://[domain]/auth/signin/microsoft-entra-id`
 1. The script hooks the IAB "url change" and "browser close" events.
 1. The internet server takes the user directly to the 3rd party sign in process, with the browser url changing to the callback url when it's done.
 1. That url is seen on the mobile, and the IAB closed. Then I extract the *session token* from the IAB cookies. That's quite a trick, but it works.


 At this point I have a session token, and the internet server has a set of auth.js cookies, and my back end database server has a record of the session.

From here on I can use the `session token` in a http authorization header to access resources on the internet server: it's an authentication token for the API. If I want, I can use that same API to grab the same sort of "session data" auth.js would normally provide.
